pkg/sql/opt/optbuilder/scope.go

```diff
@@ -726,6 +736,14 @@ func (s *scope) replaceSubquery(sub *tree.Subquery, multiRow bool, desiredColumn
 		}
 	}
 
+	if outScope.colSet().Len() != outScope.canBeProjectedColSet().Len() {
```

Comment:
I think `canBeProjectedColSet` is a misleading name, since in the normal case, we do project both `cols` and `orderByCols`. Also, the comment for `orderByCols` says that there can be overlap between it and `cols`, but then the code for `constructProjectForScope` just appends the two lists as if there is no overlap. If there can be no overlap, then this check could be simplified to just be `if len(outScope.orderByCols) != 0`. But if there can be overlap, then other locations need to be fixed. Either way, I'd add tests where there's possible overlap.

Even if you do end up allowing overlap (I'd probably try to prohibit it), then I think it'd be better to create a function like `hasExtraOrderByCols`:
```
func (s *scope) hasExtraOrderByCols() bool {
  if len(s.orderByCols) > 0 {
    cols := s.colSet()
    for _, c := range s.orderByCols {
      if !cols.contains(c) {
        return true
      }
    }
  }
  return false
}
```
This will be much faster in the common case where there are no `orderByCols`, and I think has a better name.